pkg/lib/dialogs.jsx

```diff
@@ -115,10 +153,26 @@ export const WithDialogs = ({ children }) => {
                               JSON.stringify(dialog));
             is_open.current = !!component;
             setDialog(component);
+            return new Promise((resolve, reject) => {
+                setResolvePromise(() => resolve);
```

Comment:
But it shouldn't cause a re-render right? `setDialog` already causes a re-render so these are all tied together in one state update.

Reply:
If you add strong assumptions, like

 - `setDialog()` will always run *exactly* once, even if the value doesn't change (optimization)
 - the two state updates will *always* be coalesced
 - there will *never* be a re-render in between these setState calls with previous data (which is demonstrably false, see the referenced commits)

and also are fine with abusing React state for treating synchronous internal data, then this might work. I'm just warning you, I've spent hours on debugging the previous race condition. I don't pretend I understand the precise timings of React in a detailed way, if you do and say it's fine, then go ahead.

I'll retract my -1, so that you are unblocked.

Reply:
Fwiw, I think using `setState` will probably work, but the argument for why is too complex and we should try to avoid having to argue by treating the promises like we do `is_open`. In fact, I think we can replace `is_open` with `resolvePromise != null`. But I am also ok with getting this in.